docs/development/scala.md

```diff
@@ -0,0 +1,163 @@
+
+# Scala Programming Advice
+
+- Tip: Reasonable Scala coding advice https://twitter.github.io/effectivescala/
+- Tip: You can bind names to destructuring pattern matches using the undocumented `@` syntax: 
+  ```scala
+  case class Person(name: String, age: Int)
+  val b = Person("X", 200)
+  b match {
+    case p @ Person(_, age) => println(s"Person ${p.name} age $age")
+  }
+  ```
+- Tip: Want to define a trait's method which returns a value of the type of the subclass implementing the trait?
+    Simply add a type parameter constrained to subtype traits type paramaterised by itself.
+
+  ```scala
+  trait IntrusiveListElement[T <: IntrusiveListElement[T]]:
+    def succ() : T
+    def pred() : T
+  ```
+
+## Scala Gotchas
+
+Below is an unorganised list of common pitfalls encountered when programming Scala.
+
+- Statement lists are newline-separated sequences of expressions, which simply evaluate to the last expression in the list. It is easy to accidentally silently 
+  throw away a computed value.
+  - Throwing away values in the middle of the list isn't a warning by default, but can be made one with `-Wnonunit-statement`. 
+    Unfortunately there is no standard library `ignore: _ => Unit` function, and the design of `scala.collections` in always 
+    returning non-unit, even purely side-effecting operations makes enabling this warning require somewhat more convoluted code.
+- Scala contains syntax to encourage procedural programming; if/else statements, for, while statements; 
+  the fact this is implemented through complex syntax sugar on top of functional constructs makes them surprising to both procedural and functional programmers, 
+  and best avoided. 
```

Comment:
I am a bit confused by this - is this suggesting that if/else statements and for and while loops should all be avoided? I don't think that would be useful advice at all.

Reply:
The point generally is that using procedural constructs gets you the X/Y problem that makes one reach for nonsense like `breakable`, which isn't required if you're writing functional style code.  

I don't think there's ever any reason to write a while loop in Scala they always become convoluted procedural fragments of code. Pattern matching is usually better than if/else but it does have its place. And generally I think map should be used instead of for, except for purely side-effecting operations which should be rare if you have immutable data, but they obviously mean the same thing in scala. Except that for is overloaded for do notation.

Reply:
The only thing that actually causes problems is `breakable`, which indeed should be avoided. That part was fine, which is why I didn't highlight it. 

For loops being syntactic sugar for map doesn't actually cause problems and using for can sometimes be more readable than using map. There are plenty of cases where if statements are appropriate and pattern matching is not, so more useful advice may be 'pattern matching is recommended where possible' or something along those lines? There are of course plenty of places where we have mutable data as well - for better or worse the project is not written in anything remotely close to a purely functional style and moving away from that would be far more trouble than it's worth. While loops generally only make sense when involving mutable collections but we have those.

I'm mostly concerned that advising people who work on this project (who generally have little-to-no background in functional programming) to avoid these standard procedural programming constructs that they are familiar with will just cause them to write worse, more confused code that avoids them unnecessarily.